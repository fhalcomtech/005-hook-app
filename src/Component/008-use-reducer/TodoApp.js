import React, {useEffect, useReducer} from "react";
import {useForm} from "../../hooks/useForm";
import todoReducer from "./todoReducer";

const initialState = [
    {
        id: new Date().getTime(),
        desc: "Learn react",
        done: false,
    },
];
const TodoApp = () => {
    const [todos, dispatch] = useReducer(todoReducer, initialState,);
    const [formData, handlerInputChange, resetForm] = useForm({todoDesc: ""});
    const {todoDesc} = formData;

    const handlerSubmit = (e) => {
        e.preventDefault();
        const newTodo = {
            id: new Date().getTime(),
            desc: todoDesc,
            done: false,
        }
        const action = {
            type: 'add',
            payload: newTodo
        };
        dispatch(action);
        resetForm({todoDesc: ""});
    };

    useEffect(() => {
        localStorage.setItem('todos',JSON.stringify(todos))
    }, [todos]);


    const handlerClickDelete = (todo) =>
    {
        const action = {
            type: 'delete',
            payload: todo
        };
        dispatch(action);
    }



    return (
        <>
            <h1 className="text-center">Todo App</h1>
            <hr/>
            <div className="row">
                <div className="col-7">
                    <ul className="list-group list-group-flush">
                        {todos.map(({id, desc, done}, index) => (
                            <li className="list-group-item d-flex justify-content-around" key={id}>
                                    {index + 1}. {desc}
                                <button className="btn btn-danger" onClick={(e)=>{
                                    handlerClickDelete({id,desc,done})
                                }}>delete</button>
                            </li>
                        ))}
                    </ul>
                </div>
                <div className="col-4">

                        <h2>Add Todo</h2>
                        <hr/>
                        <form onSubmit={handlerSubmit}>
                                <input
                                    type="text"
                                    className="form-control"
                                    placeholder="write todo description"
                                    name="todoDesc"
                                    value={todoDesc}
                                    onChange={handlerInputChange}
                                />
                                <button className='btn btn-success form-control mt-2 ms-0' type="submit">Add</button>
                        </form>
                    </div>
                </div>
        </>
    );
};

export default TodoApp;
