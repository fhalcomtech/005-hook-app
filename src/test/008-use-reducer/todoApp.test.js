import React from "react"
import { mount, shallow } from "enzyme"
import TodoApp from "../../Component/008-use-reducer/TodoApp"
import { initialTodoState } from "../fixtures"
import { act } from "@testing-library/react"

describe('test TodoApp component', () => { 
    const wrapper = shallow(<TodoApp/>);
    Storage.prototype.setItem = jest.fn(()=>{});
    test('TodoApp should to match snapshot', () => { 
        expect(wrapper).toMatchSnapshot();
     });

     test('TodoApp must be add a todo', () => { 
         const wapper = mount(<TodoApp/>);
        act(()=>{
            wrapper.find('TodoAdd').prop("handlerSaveTodo")(initialTodoState[0]);
            wrapper.find('TodoAdd').prop("handlerSaveTodo")(initialTodoState[1]);
        });
        expect(localStorage.setItem).toHaveBeenCalledTimes(1);
        expect(wrapper.find('h1').text().trim()).toBe('Todo App: 2')
      });

      test('TodoApp must be delete a todo', (done) => {
          wrapper.find('TodoAdd').props().handlerSaveTodo(initialTodoState[0]);
          expect(wrapper.find('h1').text().trim()).toBe('Todo App: 3');


          wrapper.find('TodoList').props().handlerClickDelete(initialTodoState[0]);
          expect(wrapper.find('h1').text().trim()).toBe('Todo App: 1');
          done();
       })
 })