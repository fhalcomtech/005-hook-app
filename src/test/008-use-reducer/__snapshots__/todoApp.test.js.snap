// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`test TodoApp component TodoApp should to match snapshot 1`] = `
<Fragment>
  <h1
    className="text-center"
  >
    Todo App: 
    0
  </h1>
  <hr />
  <div
    className="row"
  >
    <TodoList
      handlerClickDelete={[Function]}
      handlerClickDone={[Function]}
      todos={Array []}
    />
    <TodoAdd
      handlerSaveTodo={[Function]}
    />
  </div>
</Fragment>
`;
